### PR TITLE
expression: fix return type and decimal len for builtin aggregation function 'avg' and 'sum'

### PR DESCRIPTION
**Pull request by spongedu**
1. Mysql return `double` for `float` && `double` input, while TiDB return `decimal` for now.
in MySQL
```sql
mysql> create table sumavg (a int, b decimal(6,3), c double, d float);                                                                                                                                      Query OK, 0 rows affected (0.03 sec)

mysql> insert sumavg values (1, 1, 1,1);                                                                                                                                                                    Query OK, 1 row affected (0.01 sec)

mysql> select sum(c), sum(d), avg(c), avg(d) from sumavg;
Field   1:  `sum(c)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DOUBLE
Collation:  binary (63)
Length:     23
Max_length: 1
Decimals:   31
Flags:      BINARY NUM

Field   2:  `sum(d)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DOUBLE
Collation:  binary (63)
Length:     23
Max_length: 1
Decimals:   31
Flags:      BINARY NUM

Field   3:  `avg(c)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DOUBLE
Collation:  binary (63)
Length:     26
Max_length: 1
Decimals:   31
Flags:      BINARY NUM

Field   4:  `avg(d)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DOUBLE
Collation:  binary (63)
Length:     16
Max_length: 1
Decimals:   31
Flags:      BINARY NUM


+--------+--------+--------+--------+
| sum(c) | sum(d) | avg(c) | avg(d) |
+--------+--------+--------+--------+
|      1 |      1 |      1 |      1 |
+--------+--------+--------+--------+
1 row in set (0.00 sec)
```

in TiDB
```sql
mysql> create table sumavg (a int, b decimal(6,3), c double, d float);                                                                                                                                      Query OK, 0 rows affected (0.03 sec)

tidb> insert sumavg values (1, 1, 1,1);                                                                                                                                                                     Query OK, 1 row affected (0.00 sec)

tidb> select sum(c), sum(d), avg(c), avg(d) from sumavg;
Field   1:  `sum(c)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     92
Max_length: 1
Decimals:   31
Flags:      BINARY NUM

Field   2:  `sum(d)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     92
Max_length: 1
Decimals:   31
Flags:      BINARY NUM

Field   3:  `avg(c)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     92
Max_length: 6
Decimals:   31
Flags:      BINARY NUM

Field   4:  `avg(d)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     92
Max_length: 6
Decimals:   31
Flags:      BINARY NUM


+--------+--------+--------+--------+
| sum(c) | sum(d) | avg(c) | avg(d) |
+--------+--------+--------+--------+
|      1 |      1 | 1.0000 | 1.0000 |
+--------+--------+--------+--------+
1 row in set (0.00 sec)

```

2. For builtin aggregation `avg`, If return type is `decimal`,  value of `result's decimal` =  `input.decimal` + 4, while in TiDB, `result's decimal` =  `input.decimal` for now.
in MySQL:
```sql
tidb> select avg(a), avg(b) from sumavg;
Field   1:  `avg(a)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     92
Max_length: 6
Decimals:   0
Flags:      BINARY NUM 

Field   2:  `avg(b)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     92
Max_length: 9
Decimals:   3
Flags:      BINARY NUM 


+--------+-----------+
| avg(a) | avg(b)    |
+--------+-----------+
| 1.0000 | 1.0000000 |
+--------+-----------+
1 row in set (0.00 sec)
```

in TiDB:
```sql
mysql> select avg(a), avg(b) from sumavg;
Field   1:  `avg(a)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     16
Max_length: 6
Decimals:   4
Flags:      BINARY NUM 

Field   2:  `avg(b)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NEWDECIMAL
Collation:  binary (63)
Length:     12
Max_length: 9
Decimals:   7
Flags:      BINARY NUM 


+--------+-----------+
| avg(a) | avg(b)    |
+--------+-----------+
| 1.0000 | 1.0000000 |
+--------+-----------+
1 row in set (0.00 sec)
```

### CODE REVIEW

File: expression/aggregation/avg.go [hunk 1]
```diff
@@ -36,9 +36,15 @@ func (af *avgFunction) Clone() Aggregation {
 
 // GetType implements Aggregation interface.
 func (af *avgFunction) GetType() *types.FieldType {
-	ft := types.NewFieldType(mysql.TypeNewDecimal)
+	var ft *types.FieldType
+	if af.Args[0].GetType().Tp == mysql.TypeFloat || af.Args[0].GetType().Tp == mysql.TypeDouble {
```
Comment (ngaut) [inline expression/aggregation/avg.go]: DRY. Please use tp := af.Args[0].GetType().Tp, then use tp instead.

Comment (spongedu) [reply expression/aggregation/avg.go]: ok

File: expression/aggregation/sum.go [hunk 2]
```diff
@@ -44,13 +42,6 @@ func (sf *sumFunction) Update(ctx *AggEvaluateContext, sc *stmtctx.StatementCont
 
 // GetResult implements Aggregation interface.
 func (sf *sumFunction) GetResult(ctx *AggEvaluateContext) (d types.Datum) {
-	if ctx.Value.Kind() == types.KindFloat64 {
```
Comment (XuHuaiyu) [inline expression/aggregation/sum.go]: why remove this?

File: expression/aggregation/sum.go [hunk 3]
```diff
@@ -76,9 +67,13 @@ func (sf *sumFunction) CalculateDefaultValue(schema *expression.Schema, ctx cont
 
 // GetType implements Aggregation interface.
 func (sf *sumFunction) GetType() *types.FieldType {
-	ft := types.NewFieldType(mysql.TypeNewDecimal)
+	var ft *types.FieldType
+	if sf.Args[0].GetType().Tp == mysql.TypeFloat || sf.Args[0].GetType().Tp == mysql.TypeDouble {
```
Comment (XuHuaiyu) [inline expression/aggregation/sum.go]: this check can be extracted as IsTypeFloat() in types/etc.go ?

File: expression/aggregation/avg.go [hunk 4]
```diff
@@ -36,9 +36,17 @@ func (af *avgFunction) Clone() Aggregation {
 
 // GetType implements Aggregation interface.
 func (af *avgFunction) GetType() *types.FieldType {
-	ft := types.NewFieldType(mysql.TypeNewDecimal)
+	var ft *types.FieldType
+	tp := af.Args[0].GetType().Tp
+	if tp == mysql.TypeFloat || tp == mysql.TypeDouble {
```
Comment (XuHuaiyu) [inline expression/aggregation/avg.go]: ditto

File: expression/aggregation/sum.go (hunk 2 above)
Comment (spongedu) [reply expression/aggregation/sum.go]: @XuHuaiyu There should not be this branch. If `ctx.Value.Kind() == types.KindFloat64`, the return type should also be `Double` other than `NEWDECIMAL`. so just return `ctx.Value` is ok

File: expression/aggregation/sum.go (hunk 3 above)
Comment (spongedu) [reply expression/aggregation/sum.go]: I think it's ok

**Comment by XuHuaiyu** [comment]
LGTM
/run-all-tests

**Comment by coocood** [comment]
@spongedu 
Seems like this change requires updating TiKV coprocessor.

**Comment by spongedu** [comment]
@coocood yes, I see the failed tests. I'll do it later

**Comment by coocood** [comment]
LGTM

**Pull request merged**